### ISSUE TITLE
Migration class name can clash with column data types

### ISSUE THREAD

**Issue by Rans4ckeR** [dotnet/efcore#32983]
# Problem description

When adding a migration with a name that matches the data type of a column things can go wrong without an immediate indication. Only after trying to add another migration the effects become apparent and new migrations can no longer be created.

Naming a migration after an already existing type is obviously not a good idea. But not being prevented from doing so is also an issue atm. It should either be blocked, or allowed but mitigated by the generator.

# Reproduction steps

## Add a migration that clashes with a potential column data type.

```
Add-Migration DateOnly
```

The following migration is generated where e.g. the AlterColumn<> type parameter is set to the type of the migration itself since the migration class takes precedence over the type located in the System namespace. Older migrations might suddenly also start using this migration as a type since they are all in the same namespace. Same for the ModelSnapshot.

```C#
using System;
using Microsoft.EntityFrameworkCore.Migrations;

#nullable disable

namespace MigrationsNamespace
{
    public partial class DateOnly : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterColumn<DateOnly>(
                name: "ColumnName",
                table: "TableName",
                type: "date",
                nullable: true,
                oldClrType: typeof(DateTime),
                oldType: "datetime2",
                oldNullable: true);
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterColumn<DateTime>(
                name: "ColumnName",
                table: "TableName",
                type: "datetime2",
                nullable: true,
                oldClrType: typeof(DateOnly),
                oldType: "date",
                oldNullable: true);
        }
    }
}
```

So far no errors are observed, deploying and executing the migration even seems to have the intended effects.

## Add another migration

Here is where things go wrong without a real clear indication of what causes it.

```
Add-Migration NewMigration
```

This results in the following error:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Initialize(ColumnOperation columnOperation, IColumn column, RelationalTypeMapping typeMapping, Boolean isNullable, IEnumerable`1 migrationsAnnotations, Boolean inline)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.InitializeColumnHelper(ColumnOperation columnOperation, IColumn column, Boolean inline)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Diff(IColumn source, IColumn target, DiffContext diffContext)+MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.DiffCollection[T](IEnumerable`1 sources, IEnumerable`1 targets, DiffContext diffContext, Func`4 diff, Func`3 add, Func`3 remove, Func`4[] predicates)+MoveNext()
   at System.Linq.Enumerable.ConcatIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Diff(ITable source, ITable target, DiffContext diffContext)+MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.DiffCollection[T](IEnumerable`1 sources, IEnumerable`1 targets, DiffContext diffContext, Func`4 diff, Func`3 add, Func`3 remove, Func`4[] predicates)+MoveNext()
   at System.Linq.Enumerable.ConcatIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Sort(IEnumerable`1 operations, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.GetDifferences(IRelationalModel source, IRelationalModel target)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
```

# Potential solution

Blocking the use of potential conflicting types as a migration name might be difficult.
However this can be easily mitigated by generating e.g. column data types with the full namespace:

```C#
using System;
using Microsoft.EntityFrameworkCore.Migrations;

#nullable disable

namespace MigrationsNamespace
{
    public partial class DateOnly : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterColumn<System.DateOnly>(
                name: "ColumnName",
                table: "TableName",
                type: "date",
                nullable: true,
                oldClrType: typeof(System.DateTime),
                oldType: "datetime2",
                oldNullable: true);
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.AlterColumn<System.DateTime>(
                name: "ColumnName",
                table: "TableName",
                type: "datetime2",
                nullable: true,
                oldClrType: typeof(System.DateOnly),
                oldType: "date",
                oldNullable: true);
        }
    }
}
```

# Provider and version information

EF Core version: 8.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.9.0 Preview 5.0

**Comment by ajcvickers** [comment]
Duplicate of #26954
